경북대 FE 장세현 2단계 - 상품 상세 페이지 API 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
2단계 과제 - '상품 상세 페이지 API 구현하기' 내용을 PR 요청 드립니다!

---

### 과제 진행 사항

#### 목표

- 상품 상세 페이지를 구현하여 API 기능을 완성시키기

#### 요구사항

- 상품 상세 페이지
  - 참고 URL 을 참고하여 상품 상세 페이지 구현
  - API 통신의 최적화, 선언적 구조, react query 등을 활용하여 기능 구현
  - 아래 참고 API를 모두 사용하여 구현하기
    - 상품 상세 페이지에서 사용되는 API는 총 4개 ([API 명세](https://edu.nextstep.camp/s/0eoNzeZS/ls/zJw3ijiQ))
      - 상품 정보 API: /api/products/:productId
      - 상품 세부 정보 API: /api/products/:productId/detail
      - 상품 주요 리뷰 API: /api/products/:productId/highlight-review
      - 상품 관심 등록 수 API: /api/products/:productId/wish

  - 상품 관심 등록 버튼 클릭 시 낙관적 업데이트를 통해 상품 관심 등록 수를 변경할 수 있도록 구현
        (실제 API 반영은 없기 때문에, 새로고침 하면 사라지는 것이 정상)

- ErrorBoundary와 Suspense를 사용하여 코드 구조 리팩터링 하기
      (단, ErrorBoundary는 관련 라이브러리를 사용하지 않고 구현)
  => 해당 사항은 지난 Step에서 진행하였기 때문에 새로운 상품 상세 페이지에 추가 구현만 진행하였습니다.

- 본인만의 기준으로 일관된 코드 작성하기
- 기능 단위로 나누어 커밋 하기

SeHIgh committed:
docs(README): STEP 2 목표 및 과제 진행 요구사항 추가

SeHIgh committed:
feat(Product): 상품 상세 페이지 생성 및 라우팅 및 컴포넌트 추가

- pages 디렉토리 index.ts 수정

SeHIgh committed:
feat(Product): 상품 정보 및 관련 API 엔드포인트 추가

- API 엔드포인트 및 타입 정의 추가
(상품 기본 정보, 상세 정보, 하이라이트 리뷰, 찜 정보 조회)
- 상품 조회 서비스 추가

SeHIgh committed:
feat(Product): 상품 상세 페이지 및 관련 컴포넌트 구현 (상품 상세 페이지 디자인 및 상품 헤더 및 버튼 영역)

- ProductItem 컴포넌트 클릭 핸들러 이동 페이지 변경 (주문하기 페이지 -> 상품 상세 정보 페이지)
- 상품 상세 페이지 구조 및 스타일링 추가
- 상품 정보 헤더 스타일링 추가
- 상품 정보, 상세 정보, 리뷰, 찜 정보 병렬 요청
- 찜하기 및 주문하기 핸들러 추가
- 찜하기 버튼에 낙관적 업데이트를 적용

SeHIgh committed:
feat(Product): 상품 상세 페이지 구현 완료 및 README 수정

- 상품 정보 탭 구현완료

ddongule commented on README.md:
👍 상세히 작성해주셨네요!

ddongule commented on src/api/services/product.ts:
```diff
@@ -1,6 +1,14 @@
+  type ProductDetail,
+  type ProductHighlightReview,
+  type ProductWish,
+} from '@/api/types'
```
요런 방법도 있겠네용!
```suggestion
import type { RankType, TargetType, Product, ProductSummary, ProductDetail, ProductHighlightReview, ProductWish } from '@/api/types'
```

SeHIgh replied:
엇 그부분은
```
// * 타겟 타입
export enum TargetType {
  ALL = 'ALL',
  FEMALE = 'FEMALE',
  MALE = 'MALE',
  TEEN = 'TEEN',
}
// * 랭크 타입
export enum RankType {
  MANY_WISH = 'MANY_WISH',
  MANY_RECEIVE = 'MANY_RECEIVE',
  MANY_WISH_RECEIVE = 'MANY_WISH_RECEIVE',
}
```

이처럼 enum으로 정의 되어 있어서
위처럼 import가 자동 처리 되었던 것 같습니다.

RankType 이랑 TargetType을 묶어서 type으로 import 할 수는 없지만,
깔끔하게 정리하려면

```
import { RankType, TargetType } from '@/api/types'
import type { Product, ProductSummary, ProductDetail, ProductHighlightReview, ProductWish } from '@/api/types'
```

이렇게 나누는 것도 가능 할 것 같습니다!

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -0,0 +1,233 @@
+    { id: 'description' as const, label: '상품설명' },
+    { id: 'reviews' as const, label: '선물후기' },
+    { id: 'details' as const, label: '상세정보' },
+  ]
```
상수는 컴포넌트 내부에 선언되면 렌더링될때마다 새로운 tabs 배열이 생성되기 때문에 컴포넌트 외부에 선언해주시는게 죠아용!

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -0,0 +1,233 @@
+
+// * 탭 타입 정의
+type TabType = 'description' | 'reviews' | 'details'
+
```
요기두 요런 방법이 있겠네욥!
```suggestion
const PRODUCT_TABS = [
    { id: 'description', label: '상품설명' },
    { id: 'reviews', label: '선물후기' },
    { id: 'details', label: '상세정보' },
  ] as const

type TabType = typeof PRODUCT_TABS[number]['id']
```

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -0,0 +1,233 @@
+
+// * 상품 탭 컴포넌트
+export const ProductTabs = ({ productDetail, reviews }: ProductTabsProps) => {
+  const [activeTab, setActiveTab] = useState<TabType>('description')
```
```suggestion
  const [activeTab, setActiveTab] = useState<TabType>(PRODUCT_TABS[0].id)
```

ddongule replied:
요 부분들은 수정해야한다! 가 아니라 이런 방법도 있다~ 정도로 봐주시면 좋을 것 같아요!

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -0,0 +1,233 @@
+      case 'reviews':
+        return (
+          <TabContent>
+            {reviews?.reviews && reviews.reviews.length > 0 ? (
+              <ReviewsList>
+                {reviews.reviews.map((review, index: number) => (
+                  <ReviewItem key={index}>
+                    <ReviewAuthor variant="label1Bold">{review.authorName}</ReviewAuthor>
+                    <ReviewText variant="body1Regular">{review.content}</ReviewText>
+                  </ReviewItem>
+                ))}
+              </ReviewsList>
+            ) : reviews?.totalCount && reviews.totalCount > 0 ? (
+              <EmptyContent>후기 데이터를 불러오는 중...</EmptyContent>
+            ) : (
+              <EmptyContent>아직 후기가 없습니다.</EmptyContent>
+            )}
```
요부분은 조건부렌더링이 많아서 살짝 가독성이 떨어지는 느낌을 받는데, 
조건들을 선언해서 사용해서 좀 더 가독성을 높여보는건 어떨까요?

```ts
const hasReviews = reviews?.reviews && reviews.reviews.length > 0
```

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -0,0 +1,233 @@
+        {tabs.map((tab) => (
+          <TabButton
+            key={tab.id}
+            role="tab"
+            aria-selected={activeTab === tab.id}
```
접근성 측면까지 잘 고려해주셨군요! 👍

ddongule commented on src/pages/Product.tsx:
```diff
@@ -0,0 +1,260 @@
+`
+
+// * 상품명
+const ProductName = styled(Typography)``
+
+// * 상품 가격
+const ProductPrice = styled(Typography)``
+
+// * 상품 브랜드 명
+const ProductBrand = styled(Typography)``
```
아무 스타일이 없는 단순 wrapping 스타일드 컴포넌트는 불필요해보여욥!

SeHIgh replied:
저도 그렇게 생각을 해서 지우고자 했는데,
이번에 emotion을 쓰면서 전체 jsx 요소를 하나하나씩 스타일드 컴포넌트로 구성하다보니 
전체적으로 해당 태그, 요소 들이 어떤 역할을 하는지 명칭 지정을 통해 역할이 딱 한눈에 보이는 것 같아 좋은 것 같아서
이처럼 스타일에 대한 따로 지정이 없지만 해당 요소가 어떤 부분을 말하는지 알려주기 위해 위와같이 구성하게 되었습니다.

이전에 제가 공부할때는 해당 부분의 id나 class 명을 해당 부분의 역할을 명시하는 용도로 썼던 것 같은데

```
...
<Typograyphy id="ProductName" />
<Typograyphy id="ProductPrice" />
<Typograyphy id="ProductBrand" />
...
```

이렇게 사용하는 쪽이 좋을까요?

ddongule commented on src/pages/Product.tsx:
```diff
@@ -0,0 +1,260 @@
+  const queryClient = useQueryClient()
+  const productId = Number(id)
+
+  // * 병렬로 모든 상품 데이터 요청
+  const { data: productInfo } = useSuspenseQuery({
+    queryKey: ['productInfo', productId],
+    queryFn: () => fetchProductInfo(productId),
+  })
+
+  const { data: productDetail } = useSuspenseQuery({
+    queryKey: ['productDetail', productId],
+    queryFn: () => fetchProductDetail(productId),
+  })
+
+  const { data: reviews } = useSuspenseQuery({
+    queryKey: ['productReviews', productId],
+    queryFn: () => fetchProductHighlightReview(productId),
+  })
+
+  const { data: wishInfo } = useSuspenseQuery({
+    queryKey: ['productWish', productId],
+    queryFn: () => fetchProductWish(productId),
+  })
```
흠 이부분은 그냥 고민해보시면 좋을 것 같아용! 
모든 query들이 지금 useSuspenseQuery로 되어있어서 에러가 ErrorBoundary로 전달될 것으로 보여요. 그러면 개별 요청이 실패했을 경우에 사용자 피드백을 좀 다르게 주고싶다면 어떻게 주면 좋을까요?

SeHIgh replied:
저도 일단 처음에 멘토님의 말씀과 동일한 부분에 대한 문제를 인식하여
이렇게 하나의 페이지 (`Product.tsx`) 에서 전체 페치를 하는 것이 아닌
각각 개별의 데이터가 들어가는 섹션, 부분마다 컴포넌트를 나누고 파일을 분리하여서
각 컴포넌트에서 `useSuspenseQuery` 를 이용하고, `ErrorBoundary`를 각 컴포넌트 마다 따로 처리하는 방법을 처음에 생각을 했었습니다. 하지만, 이번 과제 예시 페이지 및 요구사항을 보았을 때는 그렇게 까지 구현은 불필요할 것 같다 판단 되어 이와 같이 단순화하여 구현하게 되었습니다.

또 다르게 생각해본 것은
전체를 묶어서 하나의 커스텀 훅을 구현하는데,
이때 각 데이터에 대해서는 `useQuery` 를 이용해서 받게되면,
에러 에 대한 내용을 개별로 처리가 가능하기 때문에
각 데이터에 대한 에러를 명시하고, 종합해서 전달 해주면,
각 에러에 대해서 페이지 내에서 처리하는 방안을 생각해보기도 했습니다!

ddongule commented on src/pages/Product.tsx:
```diff
@@ -0,0 +1,260 @@
+    navigate(`${ROUTE_PATH.ORDER}/${productId}`)
+  }
+
+  // * 디버깅용 데이터 로그
+  console.log('Product Detail:', productDetail)
+  console.log('Reviews:', reviews)
```
디버깅용 콘솔은 지워주시면 됴아용!

ddongule reviewed (changes requested):
안녕하세요 세현님!
구현해주신 내용 잘 살펴봤습니다. 고생하셨어요! 
전반적으로 섬세하게 작업해주신 것 같아 좋았어요.
다만 몇가지 고려해보면 좋은 점들이 보여 코멘트로 달아두었으니, 한번 확인해보시면 좋을 듯 합니당! 혹시 이해가 안가시거나 질문있으시면 언제든지 알려주세요!
조금만 더 화이팅입니당!

SeHIgh committed:
refactor: yaml 모듈 제거 및 product.ts에서 import 정리

SeHIgh committed:
refactor(ProductTabs): 탭 조건 최적화 및 상수 개별 관리

- 탭 조건을 따로 선언하고, useMemo로 최적화하여 불필요한 재계산 방지
- 탭 관련 상수를 constants 파일로 분리하여 코드 가독성 및 유지 보수성 향상 (불필요한 리렌더링도 방지)

SeHIgh committed:
fix(Product): 디버깅용 데이터 로그 제거

SeHIgh commented:
멘토님 항상 꼼꼼하고 정성스러운 리뷰를 해주셔서 정말 감사합니다.
덕분에 제 코드를 다시 한번 검토해보고, 좋은 코드를 작성하는 방법에 대해 많이 배울 수 있었습니다!

이번에는 멘토님의 리뷰 해주신 부분에 대한 코드 수정을 진행하였고, 
리뷰해주신 코멘트에 대한 답변도 조금 달아보았습니다!

---

### 코드 수정 내용

1. import 정리
  - enum과 type import를 적절히 분리하였습니다.

2. 탭 조건 최적화 및 상수 개별 관리
  - 복잡한 조건부 렌더링을 의미있는 변수명으로 분리하여 가독성을 향상 시켰습니다.
  - useMemo를 사용한 조건 변수 메모이제이션을 통해 불필요한 재계산을 방지하였습니다.
  - 탭 관련 상수를 constants로 분리하여 코드 구조를 개선시켰습니다.

3. 디버깅용 데이터 로그 제거
- production 환경에 적합하지 않은 console.log를 제거하였습니다!

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -15,19 +13,30 @@ interface ProductTabsProps {
-    { id: 'reviews' as const, label: '선물후기' },
-    { id: 'details' as const, label: '상세정보' },
-  ]
+  // * 조건 변수들 (useMemo로 최적화 - 불필요한 재계산 방지)
+  const tabConditions = useMemo(
```
👍 👍

ddongule commented on src/features/product/components/ProductTabs.tsx:
```diff
@@ -36,16 +45,16 @@ export const ProductTabs = ({ productDetail, reviews }: ProductTabsProps) => {
+            {hasReviews ? (
               <ReviewsList>
-                {reviews.reviews.map((review, index: number) => (
+                {reviews!.reviews.map((review, index: number) => (
```
실제 런타임에서 reviews가 null로 오게되면 죽을수도 있어서, non null assertion은 조심해서 사용하셔야해욥! 
아마 hasReviews로 위에서 조건부 했는데도 경고 떠서 그렇게 해주신 것 같아요 
아래처럼 사용하는 방법도 있답니다!

```suggestion
                {(reviews?.reviews ?? []).map((review, index: number) => (
```

ddongule commented on src/features/product/constants/ui.ts:
```diff
@@ -2,3 +2,13 @@
   INITIAL_SHOW_COUNT: 6,
 } as const
+
+// * 상품 탭 관련 상수
+export const PRODUCT_TAB_CONSTANTS = [
```
👍

ddongule reviewed (approved):
안녕하세요 세현님!
리뷰 반영해주신 부분 잘 봤습니다!
코멘트 한가지 더 남겨두었는데, 확인만 해주시면 될 것 같아요! 나머지는 이상 없는 것으로 보여 머지하도록 할게요 고생많으셨습니다!
화이팅이에요!
